### PR TITLE
Issue636

### LINKED ISSUE #636: unused-import false positive when assigning an element of the imported variable
Originally reported by: **BitBucket: [NeilGirdhar](http://bitbucket.org/NeilGirdhar), GitHub: @NeilGirdhar?**

---

Minimum working example:

```
#!python

from .b import x

class X:
    x = x[0]
```

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/636

### PR DESCRIPTION
**Pull request by degustaf**
### Fixes / new features
- Properly detect if imported name is assigned to same name in different scope.
Close #636

### CODE REVIEW

**Comment by rogalski** [comment]
I think this one would also solve #760, #851 (and maybe a bunch more). Seems like a nice fix to have in 2.0.

**Comment by degustaf** [comment]
It does not solve #760. But it does seem to solve #851.  Thank you @rogalski.  I'm looking for other issues this fixes.

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -895,6 +897,8 @@ def _is_variable_violation(node, name, defnode, stmt, defstmt,
             # defined in global or builtin scope
             if defframe.root().lookup(name)[1]:
                 maybee0601 = False
+                use_outer_definition = stmt == defstmt and \
```
Comment (PCManticore) [inline pylint/checkers/variables.py]: Let's not use ``\`` for line continuation, instead, let's use parentheses, as in:

```
(
    stmt == defstmt
    and ...
)
```

**Comment by degustaf** [comment]
I'm not sure why this is suddenly failing.  The failing test, and where that error "bad-reversed-sequence" is thrown, are nowhere near the code I changed.  This appears to be a regression due to PyCQA/astroid#381.

**Comment by rogalski** [comment]
@degustaf That is correct, this is a regression related to that commit, I sent PR with revert.

**Comment by rogalski** [comment]
Is it a dupe of #931?

**Comment by degustaf** [comment]
@rogalski No it's not.  The code from that bug still breaks after this patch is applied.

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -1108,6 +1116,7 @@ def visit_import(self, node):
             parts = name.split('.')
             try:
                 module = next(node.infer_name_module(parts[0]))
+
```
Comment (rogalski) [inline pylint/checkers/variables.py]: new line is not needed

File: pylint/test/functional/undefined_variable.py [hunk 3]
```diff
@@ -149,7 +149,7 @@ def test2(self, is_disabled=disabled): # [undefined-variable]
     func = lambda arg=arg: arg * arg # [undefined-variable]
 
     arg2 = 0
-    func2 = lambda arg2=arg2: arg2 * arg2
+    func3 = lambda arg2=arg2: arg2 * arg2
```
Comment (rogalski) [inline pylint/test/functional/undefined_variable.py]: change of name is not needed

File: pylint/test/functional/undefined_variable.py [hunk 4]
```diff
@@ -175,3 +175,26 @@ def test2(self, is_disabled=disabled): # [undefined-variable]
 # See https://bitbucket.org/logilab/pylint/issue/111/
 try: raise IOError(1, "a")
 except IOError as err: print(err)
+
+
+def test_conditional_comprehension():
```
Comment (rogalski) [inline pylint/test/functional/undefined_variable.py]: what exactly is tested here?

Comment (degustaf) [inline pylint/test/functional/undefined_variable.py]: What I thought was a solution added a regression where [this line](https://github.com/PyCQA/pylint/blob/dd15541d148d28d3a82639fed32f5f9f846e9832/pylint/checkers/design_analysis.py#L184) generated a pylint error.  This test verifies that no error is generated.

**Comment by degustaf** [comment]
@rogalski I addressed your issues, and added tests to confirm that this does close #848 and #900.

**Pull request merged**